feature: php8 union nullable fixer

pkly opened the pull request:
Fixes #3562

Adds a fixer which converts the php7.1 style nullables to the 8.0+ style

```php
function foo(?int $a) {}
```

becomes

```php
function foo(int|null $a) {}
```

I've played around with the idea of allowing union to nullable conversions as well but I don't understand this too well, so I decided to skip it.

Should work for properties of classes and functions, methods and so on.

pkly commented:
Ran the fixer and added some static related tests, everything is passing for me locally.
I also locked out the tests from <8.0 as the syntax is not supported anyway.

pkly commented:
Updated PR to hopefully not crash tests anymore.
I'm not quite sure what else is wrong with workflows, as the sample is not being updated on 7.4 even tho it wouldn't be, as the fixer isn't available on <8.
More or less ready to go I think, just matter of sorting out those workflow jobs.

pkly commented:
Few things I'm not 100% sure about are the priorities and interactions with other fixers, but I feel like realistically it's impossible to predict what happens there. In theory it's not necessary to have priorities for this, but it may be if there are fixers which add the nullability character.
Though I also feel like that's a weird case.

Wirone commented:
@pkly fixers' priority is important, because there are fixers that sort types in the union or align arguments in the signature - these must be run after your fixer. In terms of CI, you can run whole test suite locally too, to ensure everything passes.

I am yet going to review this, but we talked about this particular feature before and it would be good if its name was generic enough so it can cover transformations in both ways. People have different standards and I believe enforcing `?Foo` may be useful too. Did you consider implementing this in both ways?

pkly commented:
@Wirone I know about running the test suite locally, no worries. I was just editing some files online so.

I did initially consider implementing it both ways, yes. After thinking about it a bit I realized it might be a bit much for me though, as this is my first encounter with tokenization. It's been challenging enough without that, but in theory I _could_ implement that as an option (what I had in mind to begin with), though this would **only** apply to extremely simple union types (`string|null`) anyway.
I thought as the rule cannot be ran in <8.0 environments anyway it seemed kinda pointless to implement though, as the response/input types cannot be easily modified that way (only type+null allowed) and it's limiting.
About the only reason why I'd see it implemented is for BC, but even then 7.4 is EOL and for this fixer to run _at all_ you must be running in 8.0+.

Side note: I forgot about the function/method return types so I'm adding that right now.

pkly commented:
That's about it for the most basic version of this fixer.

The last things to add at this point are the priority changes related to union fixers (need to run this before) and maybe add this after any fixers which add nullability.

The `?` adding functionality is something to consider but again, seeing how PHP7 is EOL it's rather low on my priority list. That seems like more of a Rector functionality, where it'd backport code.
I understand that the syntax is not deprecated or anything but I strongly feel like it should not be used in new code.
Granted, that's my opinion, so if you have any suggestions I'll take a look later.

Wirone commented:
The fact that PHP7 is EOL does not matter because `?Foo` is totally valid syntax for PHP8 too (however I heard voices that it would be good to deprecate it). Yes, it would only work for `Foo|null` unions, but would be good to have (probably, it's not hard requirement from me, it could be added later but it would require good fixer's name in order to not break compatibility later).

pkly commented:
Why not see if there's demand for it first? I could probably add that option later if it'd be necessary.

Like I said, I'm up for renaming the fixer but I'm out of ideas on how to name it at the moment
`union_nullability` maybe? Even then it sounds a bit ambigious.

kubawerlos commented on doc/rules/language_construct/union_null.rst:
```diff
@@ -0,0 +1,61 @@
+===================
+Rule ``union_null``
```
Let's go with the name `nullable_type`, WDYT?

pkly replied:
I somewhat feel like people will be looking for union when trying to find this fixer, so it'd be good to keep that word in the name.

Wirone replied:
No, common feature of `?Foo` and `Foo|null` is nullability, not union, so if we plan to provide support for enforcing both formats (well, one of them), we need to focus on `nullable`, not on `union`. That's why `nullable_type` is better here.

pkly replied:
I mean, sure. I don't mind either way but I feel like people won't find it easily.

kubawerlos commented on tests/Fixer/LanguageConstruct/UnionNullFixerTest.php:
```diff
@@ -0,0 +1,184 @@
+    public static function provideReturnTypeCases(): iterable
+    {
+        yield [
+            '<?php function foo(): string|null {} ?>',
```
Any particular reason why not `null|string`? Wouldn't that be easier to implement?

You need to add an integration test for [`ordered_types`](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/rules/class_notation/ordered_types.rst) anyway.

pkly replied:
Personal preference I suppose. Not like putting it at the end is particularly hard, so...

I'll look into it, if you could point me to an existing example of multiple fixer interaction tests that'd help out a ton.

Wirone commented:
> Why not see if there's demand for it first? I could probably add that option later if it'd be necessary.

Yes, it can be added later, but [let's use sensible name](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6980#discussion_r1200842520) that can support both transformations.

pkly commented:
I've updated the fixer name and added a priority test, but I'm probably doing something incorrectly when it comes to the priority testing, as it seems to be failing some rules. The test itself seems to be passing though, so I think it's working as intended?

I don't really know how to more setup the priority graph stuff, as in the cookbook it's just under "ask for help", so, I am asking for help. Thank you in advance.

pkly commented:
@Wirone yeah I figured that much out but I have no idea how to update the graph since there's not even a note about it.

kubawerlos commented:
@pkly you have to add an integration test, update the graph, and both fixers descriptions, something like this: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6548/files

pkly commented:
I mean I'll look into it but honestly this part is extremely obscure and badly documented.

Wirone commented:
@pkly I agree, this is something I've learned by analysing how those tests work 🤷‍♂️. Maybe you can provide PR with docs improvements too 😉?

pkly commented on tests/Fixer/LanguageConstruct/NullableTypeFixerTest.php:
```diff
@@ -0,0 +1,184 @@
+        $this->doTest($expected, $input);
+    }
+
+    public static function providePropertyCases(): iterable
```
isn't that invalid syntax?

kubawerlos replied:
Nullable is fine: https://3v4l.org/Ot5Av - fixed version isn't. I believe the fixer needs to add a single space before null in such a case.

pkly replied:
Wow, that's cursed. I'll force it to add a whitespace if one isn't present then.

kubawerlos replied:
Let's have also `private? int$x` and `private ? int$x` to see how they are handled.

pkly commented:
Closing in favor of #7002

Wirone commented:
@pkly maybe you can contribute somehow to @paulbalandan's PR to be a co-author 🤔? Anyway, don't give up 👍.
